pkg/identity/issuerpool.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identity
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+type IssuerPool []Issuer
+
+func (p IssuerPool) Authenticate(ctx context.Context, token string) (Principal, error) {
+	url, err := extractIssuerURL(token)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, issuer := range p {
+		if issuer.Match(ctx, url) {
+			return issuer.Authenticate(ctx, token)
+		}
+	}
+	return nil, fmt.Errorf("Failed to match issuer URL %s from token with any configured providers", url)
+}
+
+func extractIssuerURL(token string) (string, error) {
+	parts := strings.Split(token, ".")
+	if len(parts) != 3 {
+		return "", fmt.Errorf("oidc: malformed jwt, expected 3 parts got %d", len(parts))
+	}
+
+	raw, err := base64.RawURLEncoding.DecodeString(parts[1])
+	if err != nil {
+		return "", fmt.Errorf("oidc: malformed jwt payload: %w", err)
+	}
+
+	var payload struct {
+		Issuer string `json:"iss"`
```

Comment:
How will this handle the issuer claim field in the config, where a custom issuer field is specified? Can we reuse the code that’s handling that in here?

Reply:
https://github.com/sigstore/fulcio/blob/f6016fda1ec6f1bd550699c7b8f53c2057ada174/pkg/oauthflow/oidc.go#L38

Reply:
Mapping the custom issuer path actually happens _post_ authentication when embedding the that customer issuer path into the issuer extension in the certificate. In the concrete example of using a Dex token, this issuer pool needs to check the issuer URL matches the _Dex_ URL here and pass the token back the _Dex_ issuer instance for authentication. The implementation of that Dex specific issuer will take the upstream issuer claim (e.g in `$.federated.connection` or what ever that claim is) and map it into the correct x509 extension (our Fulcio Issuer OID)

Reply:
So: importantly we need to use the actual `iss` claim in the issuer pool to match up with the correct issuer implementation. This will be a little more clear when I implement all those issuers again in the next PR.

Reply:
Thanks for the clarification! SG